### PR TITLE
prov/efa: info refactor part2

### PR DESCRIPTION
**Pull request by wzamazon**
keep refactoring getinfo code path.

### CODE REVIEW

File: prov/efa/src/efa_prov_info.c [hunk 1]
```diff
@@ -384,13 +371,19 @@ int efa_prov_info_alloc(struct fi_info **info,
 	if (!fi)
 		return -FI_ENOMEM;
 
-	fi->caps		= ep_dom->caps;
+	if (ep_type != FI_EP_RDM && ep_type != FI_EP_DGRAM) {
+		EFA_WARN(FI_LOG_DOMAIN, "Unsupported endpoint type: %d\n",
+			 ep_type);
```
Comment (shijin-aws) [inline prov/efa/src/efa_prov_info.c:493]: Should we fail here?

File: prov/efa/src/efa_ep.c [hunk 2]
```diff
@@ -622,68 +615,64 @@ static struct fi_ops_atomic efa_ep_atomic_ops = {
 	.compwritevalid = fi_no_atomic_compwritevalid,
 };
 
-int efa_ep_open(struct fid_domain *domain_fid, struct fi_info *info,
```
Comment (shijin-aws) [inline prov/efa/src/efa_ep.c:625]: since you modified it, could u add a docstring.

File: prov/efa/src/efa_prov_info.c [hunk 3]
```diff
@@ -64,13 +64,43 @@
 
 #define EFA_NO_DEFAULT -1
```
Comment (shijin-aws) [inline prov/efa/src/efa_prov_info.c:65]: This commit has no commit message, could you add it?

File: prov/efa/src/efa_prov_info.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/efa_prov_info.c:495]: Yes, we should. In new revision, I returned `-FI_ENODATA` here.

File: prov/efa/src/efa_ep.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/efa_ep.c:625]: I plan to refactor this function next.

I will move the rdm part of this function to `rxr_endpoing`, then rename it`efa_dgram_ep_open`. So I would prefer to  add docstring at that time.

File: prov/efa/src/efa_prov_info.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/efa_prov_info.c:65]: Added in new revision.

File: prov/efa/src/efa_prov_info.c [hunk 4]
```diff
@@ -259,118 +439,55 @@ static int efa_alloc_fid_nic(struct fi_info *fi, struct efa_device *device)
...
+		prov_info->caps			|= FI_HMEM;
+		prov_info->tx_attr->caps		|= FI_HMEM;
+		prov_info->rx_attr->caps		|= FI_HMEM;
+		prov_info->domain_attr->mr_mode	|= FI_MR_HMEM;
 	}
-
-	return 0;
 }
+#else
+void efa_prov_info_set_hmem_flags(struct fi_info *prov_info)
```
Comment (shijin-aws) [inline prov/efa/src/efa_prov_info.c:460]: Why not have the `#if` inside the function?

**Review by shijin-aws — APPROVED** [review]

File: prov/efa/src/efa_prov_info.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/efa_prov_info.c:460]: `#if` outside function is libfabric's preference.

**Pull request merged**